credmark/types/data/block_number.py

```diff
@@ -100,15 +113,52 @@ def __timestamp__(self) -> Timestamp:
             return self._timestamp
 
     def to_datetime(self):
-        return datetime.fromtimestamp(self.__timestamp__())
+        return datetime.fromtimestamp(self.__timestamp__(), tz=timezone.utc)
 
-    @ property
+    @property
     def timestamp(self) -> Timestamp:
         return self.__timestamp__()
 
-    @ property
+    @property
     def datestring(self) -> str:
         return str(self.to_datetime())
 
     # TODO: Add checking that we aren't looking into the future
     # TODO: Add BlockRange type
+
+    @classmethod
+    def from_datetime(cls, datetime_date_ts: Union[datetime, date, int, float]):
+        context = credmark.model.ModelContext.current_context
+
+        """
+        Returns the block number from the input timestamp, datetime or date.
+        For input of timestamp and datetime, the last block before the datetime is returned.
+        For input of date, the last block of the date is returned.
+        """
+        if isinstance(datetime_date_ts, int):
+            dt = datetime.fromtimestamp(datetime_date_ts, tz=timezone.utc)
+        elif isinstance(datetime_date_ts, float):
+            dt = datetime.fromtimestamp(int(datetime_date_ts), tz=timezone.utc)
+        elif isinstance(datetime_date_ts, datetime):
+            dt = datetime_date_ts.replace(tzinfo=timezone.utc)
+        elif isinstance(datetime_date_ts, date):
+            dt = datetime.combine(datetime_date_ts, datetime.max.time(), tzinfo=timezone.utc)
+        else:
+            raise ModelInputError(f'Invalid input for date/datetime/timestamp to query block_number {datetime_date_ts=}')
+        ts = int(dt.timestamp())
+
+        block_number = context.models.rpc.get_blocknumber(input=dict(timestamp=ts))['blockNumber']
```

Comment:
Just `return cls(sample_timestamp=ts)` You don't need to call the rpc, it's already being done in the `__new__` method.

However, it seems the `__new__` method is missing the max_block_number check when it looks up by timestamp. It should raise `BlockNumberOutOfRangeError` so you can fix that.